pkg/metrics/node.go

```diff
@@ -0,0 +1,194 @@
+	mgmtcontrollers "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+)
+
+const (
+	nodeProviderLabel  = "provider"
+	nodeClusterIDLabel = "cluster_id"
+)
+
+var (
+	nodeLabels = []string{nodeProviderLabel, nodeClusterIDLabel}
+	numNodes   = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: "cluster_manager",
+			Name:      "nodes",
+			Help:      "Number of nodes in ready clusters",
+		}, nodeLabels,
+	)
+	numCores = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: "cluster_manager",
+			Name:      "node_cores",
+			Help:      "Number of node cores in ready clusters",
+		}, nodeLabels,
+	)
+)
+
+const (
+	nodePollInterval = time.Duration(15 * time.Second)
+)
+
+type nodeMetrics struct {
+	nodeCache    mgmtcontrollers.NodeCache
+	clusterCache mgmtcontrollers.ClusterCache
+}
+
+func (m *nodeMetrics) collect(ctx context.Context) {
+	for range ticker.Context(ctx, nodePollInterval) {
+		logrus.Debugf("[node-metrics] collecting nodes to report metrics")
+
+		nodes, err := m.listNodes()
+		if err != nil {
+			logrus.Errorf("[node-metrics] couldn't list v3.Nodes: %v", err)
+			continue
+		}
+
+		var infos []*nodeInfo
+		for _, node := range nodes {
+			info, err := m.getNodeInfo(node)
+			if err != nil {
+				logrus.Debugf("[node-metrics] could not determine node info: %v", err)
+				continue
+			}
+			infos = append(infos, info)
+		}
+
+		setMetrics(infos)
+	}
+
+	logrus.Debugf("[node-metrics] context cancelled, exiting")
+}
+
+func (m *nodeMetrics) listNodes() ([]*v3.Node, error) {
+	var nodes []*v3.Node
+	backoff := wait.Backoff{
+		Duration: 500 * time.Millisecond,
+		Factor:   1.5,
+		Steps:    10,
+		Cap:      time.Duration(30 * time.Second),
+	}
+
+	err := wait.ExponentialBackoff(backoff, func() (done bool, err error) {
+		nodes, err = m.nodeCache.List("", labels.NewSelector())
+		if err != nil {
+			return false, nil // retry on error, may need to limit to certain types of errors
+		}
+
+		// retry if there are no nodes
+		if len(nodes) == 0 {
+			return false, nil
+		}
+
+		return true, nil
+	})
+
+	return nodes, err
+}
+
+type nodeLabelValues struct {
+	provider  string
+	clusterID string
+}
+
+type nodeInfo struct {
+	coreCount   int
+	labelValues nodeLabelValues
+}
+
+// getNodeInfo uses the v3.Node object and it's corresponding v3.Cluster object to get metric info for a node
+// if the cluster to which the node belongs is not in a Ready state, no info will be returned
+func (m *nodeMetrics) getNodeInfo(node *v3.Node) (*nodeInfo, error) {
+	cluster, err := m.clusterCache.Get(node.Namespace)
+	if err != nil {
+		return nil, fmt.Errorf("error getting cluster %s for node %s, %v", node.Namespace, node.Name, err)
+	}
+	if cluster == nil {
+		return nil, nil
+	}
+	if cluster.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if !v32.ClusterConditionReady.IsTrue(cluster) {
```

Comment:
This is something I went back and forth on, but ultimately landed on this as making the most sense.
If this wasn't here and a cluster was unreachable, we would report nodes for that cluster. Since these metrics are basically nodes under management (`cluster_manager_nodes`), I reasoned that those nodes aren't under management for cases like this.

I could see the case for considering nodes from all clusters, but then we'd likely need an additional `status` label or something to indicate whether or not the node / its cluster is active (among other statuses). This complicates the collection a bit as it's hard to nail down a set of fixed statuses for nodes from the cluster/node conditions.

Reply:
@rmweir maybe you have an opinion on this?

I could add a boolean label, something like: `is_active` that is true if the `ClusterConditionReady.IsTrue()` and `NodeConditionReady.IsTrue()` and otherwise gets false. Then we'd consider all nodes for metrics

Reply:
I added the `is_active` label